src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -0,0 +1,144 @@
+        :param end_interval: The end time of the metrics collection interval (exclusive).
+        :result: The metrics collected by this MetricCollector within the past metrics collection interval.
+        """
+        raise NotImplementedError()
+
+
+class EnvironmentMetricsService(protocol.ServerSlice):
+    """Slice for the management of metrics"""
+
+    def __init__(self) -> None:
+        super(EnvironmentMetricsService, self).__init__(SLICE_ENVIRONMENT_METRICS)
+        self.metrics_collectors: Dict[str, MetricsCollector] = {}
+        self.previous_timestamp = datetime.now()
+
+    def get_dependencies(self) -> List[str]:
+        return [SLICE_DATABASE]
+
+    def get_depended_by(self) -> List[str]:
+        return [SLICE_TRANSPORT]
+
+    async def start(self) -> None:
+        await super().start()
+        self.schedule(self.flush_metrics, COLLECTION_INTERVAL_IN_SEC, initial_delay=0, cancel_on_stop=True)
+
+    def register_metric_collector(self, metrics_collector: MetricsCollector) -> None:
+        """
+        Register the given metrics_collector.
+        """
+        if metrics_collector.get_metric_name() not in self.metrics_collectors:
+            self.metrics_collectors[metrics_collector.get_metric_name()] = metrics_collector
+        else:
+            raise Exception(f"There already is a metric collector with the name {metrics_collector.get_metric_name()}")
+
+    async def flush_metrics(self) -> None:
+        """
+        Invoked at the end of the metrics collection interval. Writes the metrics
+        collected by the MetricsCollectors in the past metrics collection interval
+        to the database.
+        """
+        now: datetime = datetime.now()
+        metric_count: List[EnvironmentMetricsCounter] = []
+        metric_non_count: List[EnvironmentMetricsNonCounter] = []
+        try:
+            for mc in self.metrics_collectors:
+                metrics_collector: MetricsCollector = self.metrics_collectors[mc]
+                metric_name: str = metrics_collector.get_metric_name()
+                metric_type: str = metrics_collector.get_metric_type()
+                metric_value: dict[str, int] = await metrics_collector.get_metric_value(self.previous_timestamp, now)
+                if metric_type == MetricType.COUNT:
+                    metric_count.append(
+                        EnvironmentMetricsCounter(metric_name=metric_name, timestamp=now, count=metric_value["count"])
+                    )
+                elif metric_type == MetricType.NON_COUNT:
+                    metric_non_count.append(
+                        EnvironmentMetricsNonCounter(
+                            metric_name=metric_name, timestamp=now, count=metric_value["count"], value=metric_value["value"]
+                        )
+                    )
+                elif metric_type == MetricType.COMPILE_RATE:
+                    raise NotImplementedError()
```

Comment:
I think this can just say that `COMPILE_RATE` is derived quantity and can not be the type of a reporter